frontend/dashboard/pages/PageLayout/PageLayout.tsx

```diff
@@ -1,29 +1,38 @@
-import AppHeader, { HeaderContext } from 'app-shared/navigation/main-header/Header';
+import AppHeader, {
+  HeaderContext,
+  SelectedContextType,
+} from 'app-shared/navigation/main-header/Header';
 import { Outlet, useNavigate } from 'react-router-dom';
+import type { NavigateFunction } from 'react-router-dom';
 import { useOrganizationsQuery } from 'dashboard/hooks/queries';
 import { useUserQuery } from 'app-shared/hooks/queries';
-import React, { useEffect, useMemo } from 'react';
+import React, { useEffect, useMemo, useState } from 'react';
 import type { IHeaderContext } from 'app-shared/navigation/main-header/Header';
+import { useTranslation } from 'react-i18next';
 
-import { userHasAccessToSelectedContext } from '../../utils/userUtils';
 import { useSelectedContext } from 'dashboard/hooks/useSelectedContext';
-import { DASHBOARD_ROOT_ROUTE } from 'app-shared/constants';
+import { typedSessionStorage } from 'app-shared/utils/webStorage';
+import { StudioPageSpinner } from '@studio/components';
+import type { Organization } from 'app-shared/types/Organization';
+import { userHasAccessToSelectedContext } from 'dashboard/utils/userUtils';
 
 export const PageLayout = () => {
+  const { t } = useTranslation();
   const { data: user } = useUserQuery();
   const { data: organizations } = useOrganizationsQuery();
 
   const selectedContext = useSelectedContext();
   const navigate = useNavigate();
+  const [isReady, setIsReady] = useState<boolean>(false);
+
+  if (selectedContext !== SelectedContextType.None) {
+    typedSessionStorage.setItem('dashboard::selectedContext', selectedContext);
+  }
 
   useEffect(() => {
-    if (
-      organizations &&
-      !userHasAccessToSelectedContext({ selectedContext, orgs: organizations })
-    ) {
-      navigate(DASHBOARD_ROOT_ROUTE);
-    }
-  }, [organizations, selectedContext, user.login, navigate]);
+    handleRedirection(selectedContext, organizations, navigate);
+    setIsReady(true);
+  }, [selectedContext, organizations, navigate]);
```

Comment:
What do you think of moving this into a custom hook? Ex. `useRedirectionGuard`? 😊 


```typescript
type UseRedirectionGuardResult = {
  isRedirectionComplete: boolean;
}
const useRedirectionGuard = (organizations: Organization[]): UseRedirectionGuardResult   => {
 const navigate = useNavigate();
 const selectedContext = useSelectedContext();
 const [isRedirectionComplete, setIsRedirectionComplete] = useState<boolean>(false);

  if (selectedContext !== SelectedContextType.None) {
    typedSessionStorage.setItem('dashboard::selectedContext', selectedContext);
  }
  
  useEffect(() => {
    handleRedirection(selectedContext, organization, navigate):
    isRedirectionComplete(true);
  }, [selectedContext, organizations, navigate]);
  
  return {
      isRedirectionComplete
  }
}
```

Then we can use the guard like this within the PageLayout.tsx:

```typescript
  const { isRedirectionComplete } = useRedirectionGuard(organizations);
  
  if (!isRedirectionComplete) return <StudioPageSpinner spinnerTitle={t('dashboard.loading')} />;
```

Reply:
I agree, it moves complexity away from the component function making it more readable. Let me know what you think of the changes.
See commit cfb6819